### PR TITLE
Feature become a host

### PR DESCRIPTION
**Pull request by f00se**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -34,8 +34,10 @@ gem 'jbuilder', '~> 2.5'
 
 gem 'bootstrap-sass'
 gem 'carrierwave'
+gem 'country_select'
```
Comment (PoslinskiNet) [inline Gemfile]: No version :) ?

File: Gemfile [hunk 2]
```diff
@@ -34,8 +34,10 @@ gem 'jbuilder', '~> 2.5'
...
 gem 'decent_exposure'
 gem 'devise'
+gem 'enum_help'
```
Comment (PoslinskiNet) [inline Gemfile]: Why you need this gem?

File: app/controllers/places_controller.rb [hunk 3]
```diff
@@ -1,8 +1,26 @@
 class PlacesController < ApplicationController
-  expose!(:q, -> { Place.ransack(params[:q]) })
+  before_action :authenticate_user!, except: %i[show index]
```
Comment (PoslinskiNet) [inline app/controllers/places_controller.rb]: Consider move this logic to RestrictedController

File: app/controllers/places_controller.rb [hunk 4]
```diff
@@ -1,8 +1,26 @@
...
   expose(:places, -> { q.result(distinct: true).page(params[:page]) })
   expose(:place)
 
-  def show; end
+  def create
+    redirect_to edit_place_path(current_user.places.create(place_params))
+  end
+
+  def update
+    Place.find(params[:id]).update_attributes!(place_params)
```
Comment (PoslinskiNet) [inline app/controllers/places_controller.rb]: Without validation?

File: app/models/advantage.rb [hunk 5]
```diff
@@ -0,0 +1,3 @@
+class Advantage < ApplicationRecord
+  has_and_belongs_to_many :places
```
Comment (PoslinskiNet) [inline app/models/advantage.rb]: Avoid this helper method as a fire. Use explicit version instead. It reduces customization a lot.

File: app/models/area.rb [hunk 6]
```diff
@@ -0,0 +1,3 @@
+class Area < ApplicationRecord
+  has_and_belongs_to_many :places
```
Comment (PoslinskiNet) [inline app/models/area.rb]: Avoid this helper method as a fire. Use explicit version instead. It reduces customization a lot.

File: app/models/place.rb [hunk 7]
```diff
@@ -1,5 +1,21 @@
...
+  accepts_nested_attributes_for :areas
+
+  has_and_belongs_to_many :advantages
+  accepts_nested_attributes_for :advantages
+
+  scope(:actives, -> { where(active: true) })
 
   default_scope { order(created_at: :desc) }
+
+  enum flat_type: %w[entire_room private_room shared_room]
```
Comment (PoslinskiNet) [inline app/models/place.rb]: helpers, scopes, relations, enums, validations - keep that order and group it together for better readability.

File: config/routes.rb [hunk 8]
```diff
@@ -2,6 +2,7 @@
   devise_for :users, controllers: { omniauth_callbacks: "users/omniauth_callbacks" }
 
   resources :places
```
Comment (PoslinskiNet) [inline config/routes.rb]: Are you using all actions already?

File: spec/controllers/places_controller_spec.rb [hunk 9]
```diff
@@ -35,6 +35,45 @@
...
     end
   end
+
+  describe '#create' do
+    let(:user) { create :user }
+    let(:place_params) { attributes_for(:place) }
+    let(:call_request) { post :create, params: { place: place_params } }
+
+    it 'create new place and redirect to edit page' do
+      sign_in user
```
Comment (PoslinskiNet) [inline spec/controllers/places_controller_spec.rb]: Seems that this part is repeated all the time - you could move it to common block for specs.

File: spec/factories/place.rb [hunk 10]
```diff
@@ -1,8 +1,25 @@
 FactoryGirl.define do
   factory :place do
-    title { Faker::Lorem.paragraph(1) }
+    title { Faker::Lorem.sentence }
+    active true
+    description { Faker::Lorem.sentences(10).join(' ') }
+    price { Faker::Base.numerify('##') }
+    city { Faker::Address.city }
+    street { Faker::Address.street_name }
```
Comment (PoslinskiNet) [inline spec/factories/place.rb]: All of those properties are required? If not, then factory should only produce on default - the minimum required a valid object. If you want to have some extra data in the factory take a look at traits.

File: Gemfile (hunk 1 above)
Comment (f00se) [inline Gemfile]: Everything have no version - something went wrong with earlier rebases

File: Gemfile (hunk 2 above)
Comment (f00se) [inline Gemfile]: To print better enums as collections in _place_form.html.haml

**Comment by PoslinskiNet** [comment]
🔴 specs

File: app/controllers/places_controller.rb (hunk 4 above)
Comment (f00se) [inline app/controllers/places_controller.rb]: I saw something similar in web... I will fix it.

File: app/models/advantage.rb (hunk 5 above)
Comment (f00se) [inline app/models/advantage.rb]: Ok, I will change that.

File: config/routes.rb (hunk 8 above)
Comment (f00se) [inline config/routes.rb]: Nearly all - without `delete`. So for one method it's necessary to write exception?

**Comment by f00se** [comment]
Feedback applied, but without better tests. Tomorrow I fix that. Deadline is coming...

File: spec/seeds/seeds_spec.rb [hunk 11]
```diff
@@ -0,0 +1,46 @@
+# require 'rails_helper'
```
Comment (PoslinskiNet) [inline spec/seeds/seeds_spec.rb]: Hmm :) ? What is the purpose of that?

File: test/models/amenities_place_test.rb [hunk 12]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class AmenitiesPlaceTest < ActiveSupport::TestCase
+  # test "the truth" do
```
Comment (PoslinskiNet) [inline test/models/amenities_place_test.rb]: Empy?

File: app/policies/place_policy.rb [hunk 13]
```diff
@@ -0,0 +1,36 @@
+class PlacePolicy
+  attr_reader :user, :place
+
+  def initialize(user, place)
+    @user = user
+    @place = place
+  end
+
+  def edit?
+    (place.user == user) || user.has_role?(:admin)
```
Comment (PoslinskiNet) [inline app/policies/place_policy.rb]: You can extract this to private method that will be reused. Also this method can give a meaning to the conditional. For example place_owner? || admin?.

File: app/policies/place_policy.rb [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+  def update?
+    (place.user == user) || user.has_role?(:admin)
+  end
+
+  def destroy?
+    (place.user == user) || user.has_role?(:admin)
+  end
+
+  def create?
+    !user.nil?
```
Comment (PoslinskiNet) [inline app/policies/place_policy.rb]: Why?

File: app/policies/place_policy.rb (hunk 13 above)
Comment (f00se) [inline app/policies/place_policy.rb]: Changed in next PR - #12

File: app/policies/place_policy.rb (hunk 14 above)
Comment (f00se) [inline app/policies/place_policy.rb]: I was trying to wrote policy for not logged in user - guest. So if user is nil, this policy don't run anyway.  It's my mistake - should be `user`.

**Pull request merged**